### ISSUE TITLE
[BUG] Profiling code increases performance

### ISSUE THREAD

**Issue by rhshadrach** [cython/cython#5821]
### Describe the bug

I'm working on upgrading to Cython 3.x in pandas (https://github.com/pandas-dev/pandas/pull/55179), looking into an ASV regression `series_methods.Fillna.time_ffill('Float64')`. Compiling locally using Cython 3.0.5 I am able to reproduce the timing reported in the OP there (~2.7ms). So I go to profile it by adding `# cython: profile=True` to the top of `algos.pyx`, recompile, and now the same code takes ~1.2ms.

I'm guessing this isn't expected?

### Code to reproduce the behaviour:

```python
import numpy as np
import pandas as pd

ser = Series(np.random.randn(10**6), dtype='Float64')
ser[::2] = np.nan
%timeit ser.ffill()
```

### OS

Linux

### Python version

3.10.12

### Cython version

3.0.5

### Additional context

Cython method:

https://github.com/pandas-dev/pandas/blob/55d78caa38925e8cf94623adcd0721cc15a56bdd/pandas/_libs/algos.pyx#L628

**Comment by WillAyd** [comment]
Does the timing change if you use a seed? Maybe the timing changes are value dependent?

**Comment by rhshadrach** [comment]
No - I must have run this a dozen times because I didn't believe the results. Here it is without any randomness. This was run on the main branch of pandas with Cython 3.0.5 to compile.

```
ser = Series(np.zeros(10**6), dtype='Float64')
ser[::2] = np.nan
for _ in range(5):
    %timeit ser.ffill()

# Without profiling
# 2.5 ms ± 37.1 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 2.49 ms ± 23 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 2.48 ms ± 20.9 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 2.48 ms ± 22 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)
# 2.5 ms ± 55.3 µs per loop (mean ± std. dev. of 7 runs, 100 loops each)

# With profiling
# 1.2 ms ± 5.77 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# 1.19 ms ± 7.12 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# 1.2 ms ± 8.97 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# 1.2 ms ± 7.41 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
# 1.19 ms ± 5.85 µs per loop (mean ± std. dev. of 7 runs, 1,000 loops each)
```

**Comment by WillAyd** [comment]
Hmm OK. And these are all built with the same optimization level right? If using conda I've been bitten too many times to count by how it makes changes to CFLAGS which can impact performance.

**Comment by rhshadrach** [comment]
> And these are all built with the same optimization level right?

Yes.

**Comment by da-woods** [comment]
I get a smaller effect (~10% speedup) but I can reproduce this I think.

It's not completely implausible to me -  it'll likely be to do with acquiring the GIL either conditionally or unconditionally at the start of functions.

**Comment by rhshadrach** [comment]
Thanks @da-woods 

> it'll likely be to do with acquiring the GIL either conditionally or unconditionally at the start of functions.

In the case of `pad_inplace`, we don't release the GIL at all. I assume that means we're acquiring the GIL unconditionally, and I'd assume the same happens when profiling. So I'm not sure I understand how the performance differs.

**Comment by da-woods** [comment]
Hmmm - I'm definitely puzzled.

Conclusions:

* Using `perf` tells me that the almost all the time is in `pad_inplace` (and the rest is just in `__memmove_avx_unaligned_erms`. Certainly the difference between the runs is is `pad_inplace`.  (possibly you'd already verified this).
* Nothing to do with the GIL - as you say, you're not releasing it
* There's nothing odd in the code generation. Tracing adds:

    ```
    __Pyx_TraceDeclarations
    [...]
    __Pyx_TraceCall("__pyx_fuse_9pad_inplace", __pyx_f[0], 628, 0, __PYX_ERR(0, 628, __pyx_L1_error));
    [...]
    __Pyx_TraceReturn(__pyx_r, 0);
    ```

    and all of those should just add a small amount of work.
* Something about the Pandas build system means I can't get `perf` to give me C source output
* I seem to get similar results (~10% speed-up with profiling) on both Python 3.10 and 3.11, so nothing interesting there.
* I can't reproduce it when I extract `pad_inplace` out of Pandas (together with the two other functions it uses) - the timings then end up basically the same. However I can get annotated C source.
* Almost all the time I spent in the

    ```
    for i in range(N):
        if mask[i]:
    ```

    loop (as you would expect). This appears to be the case for both the profiling/non-profiling versions in Pandas.
* In Pandas the assembly generated for these loops is slightly different (although the C source is identical):

    ```
    #### with profiling ###################
    Percent│870:┌─→cmp        %r8d,%ecx
           │    │↓ jge        87f
      3.79 │    │  movsd      %xmm0,(%rdx)
      1.63 │    │  add        $0x1,%ecx
     10.59 │    │  mov        %sil,(%rax)
           │87f:│  add        $0x1,%rax
      2.09 │    │  add        %r14,%rdx
      1.86 │    │  cmp        %rax,%rbp
           │    │↓ je         8a4
     21.74 │88b:├──cmpb       $0x0,(%rax)
     21.97 │    └──jne        870
      0.15 │       add        $0x1,%rax
     21.17 │       movsd      (%rdx),%xmm0
      1.94 │       xor        %esi,%esi
      9.91 │       xor        %ecx,%ecx
           │       add        %r14,%rdx
           │       cmp        %rax,%rbp
      3.02 │     ↑ jne        88b
    
    #### without profiling ###############
      0.12 │830:   cmp        %r8d,%ecx                                                                                                                          
           │     ↓ jge        83f                                                                                                                                
     11.75 │       movsd      %xmm0,(%rdx)                                                                                                                      
      0.21 │       add        $0x1,%ecx                                                                                                                         
      8.78 │       mov        %sil,(%rax)                                                                                                                       
      0.79 │83f:   add        $0x1,%rax                                                                                                                         
      3.64 │       add        %r14,%rdx                                                                                                                         
      1.48 │       cmp        %rax,%rbp                                                                                                                         
           │     ↑ je         680                                                                                                                               
     35.71 │84f:   cmpb       $0x0,(%rax)                                                                                                                       
     11.27 │     ↑ jne        830                                                                                                                               
     15.94 │       movsd      (%rdx),%xmm0                                                                                                                      
      0.19 │       xor        %esi,%esi                                                                                                                          
      0.64 │       xor        %ecx,%ecx                                                                                                                          
      9.31 │     ↑ jmp        83f                                                                                                                                
           │       xchg       %ax,%ax                                                                                                                            
           │860:   mov        %rdx,%rdi      
    ```
    basically just a couple of extra `add`s have appeared.

* For the version I've taken out of Pandas the assembly is the same with/without profiling (but isn't an exact match for either of the versions above).

So basically, I think this is the C compiler doing something odd. There's no relevant changes in the generated c code.

**Comment by WillAyd** [comment]
> * Something about the Pandas build system means I can't get `perf` to give me C source output

I assume you need a debug build to do this but our build system generates a release build by default. You might want to try:

```
pip install -ve . --no-build-isolation --config-settings=builddir="debug" --config-settings=setup-args="-Dbuildtype=debug"
```

https://pandas.pydata.org/docs/development/debugging_extensions.html

**Comment by da-woods** [comment]
> I assume you need a debug build to do this but our build system generates a release build by default.

I think really I'd want "release with debug symbols", rather than a pure debug build. I did try adding `-g` to the compile flags in setup.py, but that possibly wasn't enough.

For now I think I've reached the end of what I can usefully learn here anyway.

**Comment by scoder** [comment]
Thanks for investigating this so deeply, @da-woods.

> I think this is the C compiler doing something odd

It might end up applying different optimisation heuristics when it sees the additional profiling calls. This code might just hit an unlucky tipping point where the default heuristics don't work well any more but slight modifications trigger a switch in heuristics that work visibly better.

To me, this suggests that enabling PGO might help the C compiler to take better decisions also in the non-profiling case.

**Comment by WillAyd** [comment]
@da-woods makes sense. If you still cared you could also try "debugoptimized" as the build type argument in the above command. The default build system in pandas is meson so setup.py is not used by default, just there for legacy reasons

Noted that still may not matter just sharing as an FYI. Thank you for the wonderful investigative work!

**Comment by scoder** [comment]
I'm closing this ticket since I don't see Cython doing anything wrong here. It seems to be an issue with C compiler optimisation. PGO, i.e. a better informed C compiler optimiser, would likely help.
